### ISSUE TITLE
outq_len() returns 1 for several minutes before it becomes to return 0

### ISSUE THREAD

**Issue by muumu** [edenhill/librdkafka#877]
Description
===========
I am using RdKafka::Producer to post messages to my Kafka broker cluster.
When I wait for outq_len() to be 0 after I produce 100 messages, sometimes outq_len() returns 1 for several minutes and then returns 0.
I think it is strange that outq_len() returns just 1.
Do you have any idea about the cause or the way to specify it?
Some errors in Kafka brokers I excerpts below may be related to this issue.
I am using high-level KafkaConsumer in my consumer servers and the issue solved by shutting down the consumers, so I suspect that something related to the errors between brokers and consumers might cause this issue.

How to reproduce
================
It is not easy to reproduce the issue because it happens occasionally by unknown cause.

Checklist
=========
Please provide the following information:

 - [x] librdkafka version (release number or git tag): 0.9.2rc
 - [x] Apache Kafka version: 0.10.0.1
 - [ ] librdkafka client configuration:
 - [x] Operating system: CentOS 6
 - [ ] Using the legacy Consumer
 - [x] Using the high-level KafkaConsumer
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [x] Provide broker log excerpts
 - [ ] Critical issue

```
[2016-10-06 10:37:10,507] ERROR Closing socket for 100.98.114.220:9092-100.98.120.193:50498 because of error (kafka.network.Processor)
kafka.network.InvalidRequestException: Error getting request for apiKey: -12682 and apiVersion: -1
    at kafka.network.RequestChannel$Request.liftedTree2$1(RequestChannel.scala:95)
    at kafka.network.RequestChannel$Request.<init>(RequestChannel.scala:87)
    at kafka.network.Processor$$anonfun$processCompletedReceives$1.apply(SocketServer.scala:488)
    at kafka.network.Processor$$anonfun$processCompletedReceives$1.apply(SocketServer.scala:483)
    at scala.collection.Iterator$class.foreach(Iterator.scala:893)
    at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
    at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
    at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
    at kafka.network.Processor.processCompletedReceives(SocketServer.scala:483)
    at kafka.network.Processor.run(SocketServer.scala:413)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Unexpected ApiKeys id `-12682`, it should be between `0` and `18` (inclusive)
    at org.apache.kafka.common.protocol.ApiKeys.forId(ApiKeys.java:71)
    at org.apache.kafka.common.requests.AbstractRequest.getRequest(AbstractRequest.java:39)
    at kafka.network.RequestChannel$Request.liftedTree2$1(RequestChannel.scala:92)
    ... 10 more
```
100.98.114.220: IP address of one of brokers
100.98.120.193: IP address of one of consumers

**Comment by edenhill** [comment]
That InvalidRequestException looks bad.
Was there anything logged by the consumer around the time that exception happened?
Can you provide your consumer and producer config?


There shouldn't be any correlation between consumers and producers in Kafka, so it sounds weird that shutting down the consumers would unblock the producers.

**Comment by muumu** [comment]
Thank you for your quick response.
I have checked logs from 10:37:07 to 10:37:13 of the consumer that appeared in the above kafka broker log at 10:37:10, but I found no error.
The consumer uses customized RdKafka::EventCb that logs EVENT_ERROR, EVENT_STATS, EVENT_LOG, EVENT_THROTTLE, and the other events (default case in switch), so consumer should have logged if it had received any events from the brokers.

Both of producer and consumer configurations are default settings except group.id, broker.list and event_cb.
```
builtin.features = gzip,snappy,ssl,sasl,regex
client.id = rdkafka
message.max.bytes = 1000000
receive.message.max.bytes = 100000000
max.in.flight.requests.per.connection = 1000000
metadata.request.timeout.ms = 60000
topic.metadata.refresh.interval.ms = 300000
topic.metadata.refresh.fast.cnt = 10
topic.metadata.refresh.fast.interval.ms = 250
topic.metadata.refresh.sparse = true
socket.timeout.ms = 60000
socket.blocking.max.ms = 100
socket.send.buffer.bytes = 0
socket.receive.buffer.bytes = 0
socket.keepalive.enable = false
socket.max.fails = 3
broker.address.ttl = 1000
broker.address.family = any
reconnect.backoff.jitter.ms = 500
statistics.interval.ms = 0
enabled_events = 0
log_cb = 0x7f01392a7890
log_level = 6
log.thread.name = false
log.connection.close = true
socket_cb = 0x7f01392b1410
open_cb = 0x7f01392c52e0
internal.termination.signal = 0
api.version.request = false
api.version.fallback.ms = 1200000
broker.version.fallback = 0.9.0
security.protocol = plaintext
sasl.mechanisms = GSSAPI
sasl.kerberos.service.name = kafka
sasl.kerberos.principal = kafkaclient
sasl.kerberos.kinit.cmd = kinit -S "%{sasl.kerberos.service.name}/%{broker.name}" -k -t "%{sasl.kerberos.keytab}" %{sasl.kerberos.principal}
sasl.kerberos.min.time.before.relogin = 60000
group.id = 0
partition.assignment.strategy = range,roundrobin
session.timeout.ms = 30000
heartbeat.interval.ms = 1000
group.protocol.type = consumer
coordinator.query.interval.ms = 600000
enable.auto.commit = false
auto.commit.interval.ms = 0
enable.auto.offset.store = true
queued.min.messages = 100000
queued.max.messages.kbytes = 1000000
fetch.wait.max.ms = 100
fetch.message.max.bytes = 1048576
fetch.min.bytes = 1
fetch.error.backoff.ms = 500
offset.store.method = broker
enable.partition.eof = true
queue.buffering.max.messages = 100000
queue.buffering.max.kbytes = 4000000
queue.buffering.max.ms = 1000
message.send.max.retries = 2
retry.backoff.ms = 100
compression.codec = none
batch.num.messages = 1000
delivery.report.only.error = false
```
I will try to reproduce this issue and to find out the cause.
Please inform me if you get any other ideas. Thank you.

**Comment by muumu** [comment]
I have found out the cause of broker error (kafka.network.InvalidRequestException).
I did not set api.version.request param, so the solution is setting it true.
```
props["api.version.request"] = "true";
```
I overlooked the following requirements.
https://github.com/edenhill/librdkafka/wiki/Broker-version-compatibility

This solution also suppressed the out queue blocking, which seems weird as you said though.
